### PR TITLE
Windows support on MSYS2, stubs unimplemented unused stuff when on WIN32

### PR DESCRIPTION
**Pull request by franciscod**
**heavily** based on @mappu guide on #9. This aims to close that issue, and enables openage to be built on MSYS2 with traditional `cmake` by stubbing not yet implemented (and coincidentally, unused) features under WIN32. 

A better solution is to implement the features (I think this is desirable on os.cpp functions), or strip the entire feature from all platforms if it's not useful (I never used the console tests, they seem dated).

Needs #218 so it should be merged first, and this branch should be rebased. If you want to test, just pull TheJJ/math-constants into master, and then this branch into the result, it should build.

In contrast with mappu's guide, this doesn't modify`./configure` nor `buildsystem/simple`. Also, the majority of the patches that mappu required are no longer necessary due to better code quality :) Oh, and we now have opusfile on the MSYS2 repos, so no manual compiling of that dep.

TL;DR: 
```
# convert assets or copy them already converted
mkdir build && cd build
cmake -G "MSYS Makefiles" .. #configures flawlessly
make #compiles flawlessly
cd .. #changes directory flawlessly
./build/cpp/openage #runs flawlessly
```

---

This was on a Windows7 32bit VM. This guide should work without issues, please try it out and comment about your experience!

- Install [msys2](https://msys2.github.io/) on C:\msys2. When the wizard asks, don't open the shell. 

`/` means C:\msys2 and `~` is `/home/<username>`.

- Move AoE2 directory to ~, assuming "~/age2/Data" is the right directory (contains a lot of dat an drs).

From now on, you need to use mingw32_shell.bat from C:\msys2. Open up a shell and go:
- `pacman -Syu`
- copypasta this package installing havoc:
```
pacman -S git mingw-w64-i686-{gcc,pkg-config,make,openssl,libogg,opus{,file,-tools},SDL2{,_image},glew,freetype,ftgl,python3{,-{,numpy,Pillow}}}
```
- copy mingw32-make to make, so standard tools don't whine:
```
cp /mingw32/bin/{mingw32-,}make.exe
```
- get the windows branch from my repo:
```
cd ~
git clone https://github.com/franciscod/openage
cd openage
git checkout windows
```
- cmake magic!:
```
mkdir build && cd build
cmake -G "MSYS Makefiles" ..
make
cd ..
```
- you might need to install dejavu serif from http://sourceforge.net/projects/dejavu/files/dejavu/2.34/dejavu-fonts-ttf-2.34.zip
- convert *dat ass*ets:
```
PYTHONPATH=py python3 -m openage.convert -v media -o "assets/converted" "/c/Program Files (x86)/Age of Empires/" graphics:*.* terrain:*.* sounds0:*.* sounds1:*.* gamedata0:*.* gamedata1:*.* gamedata2:*.* interface:*.*
```
- Convert all *.docx files from CRLF to LF: 
```
find ./assets/converted/ | grep 'docx$' | xargs -n1 sed -rie 's/\r\n/\n/'
```
- enjoy openage on MSYS2! :dancer: 

```
./build/cpp/openage
```

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -295,8 +295,8 @@ def forcesymlink(linkto, name):
     os.symlink(linkto, name)
 
 # create the build dir and symlink it to 'bin'
-forcesymlink(bindir, 'bin')
-forcesymlink('%s/cpp/openage' % bindir, 'openage')
+# forcesymlink(bindir, 'bin')
+# forcesymlink('%s/cpp/openage' % bindir, 'openage')
```
Comment (franciscod) [inline configure]: I don't have a clue on this one. Wat do?

**Comment by franciscod** [comment]
What should we do about symlinks?

File: buildsystem/simple [hunk 2]
```diff
@@ -15,9 +15,10 @@
 PYVER=3.4m
 PYVERDOTLESS=34m
 shopt -s globstar
+set -e
```
Comment (franciscod) [inline buildsystem/simple]: I think this one should stay: we don't want to keep building if a previous step failed.

File: buildsystem/simple [hunk 3]
```diff
@@ -15,9 +15,10 @@
...
 
 # build py ext module (openage.convert.cabextract.lzxd)
-g++ -O3 -Wall -Wextra -pedantic -std=c++11 -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -fwrapv -fstack-protector-strong --param=ssp-buffer-size=4 -fPIC -I/usr/include/python${PYVER} -lpthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro -lpython${PYVER} -o py/openage/convert/cabextract/lzxd.cpython-${PYVERDOTLESS}.so py/openage/convert/cabextract/lzxd/{lzxd.cpp,pyinterface.cpp,util.cpp}
+g++ -O3 -Wall -Wextra -pedantic -std=c++11 -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -fwrapv -fstack-protector-strong --param=ssp-buffer-size=4 -I/mingw32/include/python${PYVER} -lpthread -shared -Wl,-O1,--sort-common,--as-needed -Wno-format -o py/openage/convert/cabextract/lzxd.cpython-${PYVERDOTLESS}.so py/openage/convert/cabextract/lzxd/{lzxd.cpp,pyinterface.cpp,util.cpp} -lpython${PYVER}
```
Comment (franciscod) [inline buildsystem/simple]: these changes should be addressed by CMake

File: configure (hunk 1 above)
Comment (franciscod) [inline configure]: Should we set bindir to 'bin' directly?

File: configure [hunk 4]
```diff
@@ -330,6 +330,8 @@ for raw_cmake_arg in args.raw_cmake_args:
 
 invocation.append('--')
 invocation.append(project_root)
+invocation.append("-G")
+invocation.append("MSYS Makefiles")
```
Comment (franciscod) [inline configure]: How should we say to CMake that it should use this generator, but just on WIN32 platform?

File: buildsystem/simple (hunk 3 above)
Comment (franciscod) [inline buildsystem/simple]: whoops, except -Wno-format

**Comment by franciscod** [comment]
Submitted a PR so we can get opusfile on mingw via pacman :dancer: https://github.com/Alexpux/MINGW-packages/pull/445

**Comment by franciscod** [comment]
btw look at dis cmake buildsystem :open_mouth: https://github.com/Unvanquished/Unvanquished

**Comment by franciscod** [comment]
it was merged! msys2 repos now have opusfile! :D

https://github.com/Alexpux/MINGW-packages/pull/445

```
$ pacman -Ss opusfile
mingw32/mingw-w64-i686-opusfile 0.6-1 [installed]
    Library for opening, seeking, and decoding .opus files (mingw-w64)
mingw64/mingw-w64-x86_64-opusfile 0.6-1
    Library for opening, seeking, and decoding .opus files (mingw-w64)
```

**Comment by franciscod** [comment]
opusfile on aur :3 https://aur.archlinux.org/packages/mingw-w64-opusfile

**Comment by franciscod** [comment]
Recap: this was ~~about patching and tweaking the buildsystem and running and compiling instructions and...~~ a big mess, so I splitted into the following:
- #233 for the buildsystem and running, it's a take on Linux+mingw-w64+Wine instead of MSYS2+Windows7. It runs!
- #234 for the patches stubbing currently unused things on WIN32.

Thanks!
